internal/step/plugin/provider.go

```diff
@@ -922,52 +1080,88 @@ func (r *runningStep) closeComponents(closeATP bool) error {
 	return nil
 }
 
+// Note: Caller must add 1 to the waitgroup before calling.
 func (r *runningStep) run() {
-	r.wg.Add(1) // Wait for the run to finish before closing.
 	defer func() {
 		r.cancel()  // Close before WaitGroup done
 		r.wg.Done() // Done. Close may now exit.
 	}()
-	container, err := r.deployStage()
-	if err != nil {
-		r.deployFailed(err)
+	pluginConnection := r.startPlugin()
+	if pluginConnection == nil {
 		return
 	}
+	defer func() {
+		err := pluginConnection.Close()
+		if err != nil {
+			r.logger.Errorf("failed to close deployed container for step %s/%s", r.runID, r.pluginStepID)
+		}
+	}()
+	r.postDeployment(pluginConnection)
+}
+
+// Deploys the plugin, and handles failure cases.
+func (r *runningStep) startPlugin() deployer.Plugin {
+	pluginConnection, contextDoneEarly, err := r.deployStage()
+	if contextDoneEarly {
+		if err != nil {
+			r.logger.Debugf("error due to step early closure: %s", err.Error())
+		}
+		r.closedEarly(StageIDEnabling, true)
+		return nil
+	} else if err != nil {
+		r.deployFailed(err)
+		return nil
+	}
 	r.lock.Lock()
 	select {
 	case <-r.ctx.Done():
-		if err := container.Close(); err != nil {
-			r.logger.Warningf("failed to remove deployed container for step %s/%s", r.runID, r.pluginStepID)
+		if err := pluginConnection.Close(); err != nil {
+			r.logger.Warningf("failed to close deployed container for step %s/%s", r.runID, r.pluginStepID)
 		}
 		r.lock.Unlock()
-		r.transitionToCancelled()
-		return
+		r.closedEarly(StageIDEnabling, false)
+		return nil
 	default:
-		r.container = container
+		r.container = pluginConnection
 	}
 	r.lock.Unlock()
-	r.logger.Debugf("Successfully deployed container with ID '%s' for step %s/%s", container.ID(), r.runID, r.pluginStepID)
+	r.logger.Debugf("Successfully deployed container with ID '%s' for step %s/%s", pluginConnection.ID(), r.runID, r.pluginStepID)
+	return pluginConnection
+}
 
+func (r *runningStep) postDeployment(pluginConnection deployer.Plugin) {
 	r.logger.Debugf("Checking to see if step %s/%s is enabled", r.runID, r.pluginStepID)
-	enabled, err := r.enableStage()
-	if err != nil {
-		r.startFailed(err)
+	enabled, contextDoneEarly := r.enableStage()
+	if contextDoneEarly {
+		r.closedEarly(StageIDStarting, true)
+		return
 	}
 	r.logger.Debugf("Step %s/%s enablement state: %t", r.runID, r.pluginStepID, enabled)
 	if !enabled {
 		r.transitionToDisabled()
 		return
 	}
-	if err := r.startStage(container); err != nil {
+
+	var forceCloseTimeoutMS int64
+	var err error
+	if contextDoneEarly, forceCloseTimeoutMS, err = r.startStage(pluginConnection); contextDoneEarly {
```

Comment:
Do I understand this syntax correctly? The `var` statements outside the `if` scope are extending the unpacked `forceCloseTimeoutMS` and `err` lifetimes beyond the `if`? Huh.

(What initially confused me here was that GitHub is sometimes very "helpful" in giving links to variable definitions ... but in this case `startStage`'s link took me to the *previous* definition, with only one return value, which really confounded me until I found the right one. Sometimes being helpful is *not* helpful ... 🐞 )